wallet: Do not iterate a directory if having an error while accessing it

hebasto opened the pull request:
On Windows when `ListDatabases` tries to iterate any system folder, e.g., "System Volume Information", it falls into an infinite loop.

This PR fixes this bug. Now the `debug.log` contains:
```
2021-05-12T09:07:53Z ListDatabases: Access is denied D:/System Volume Information -- skipping.
```

An easy way to reproduce the bug and test this PR is to pass the `-walletdir=D:\` command-line option, and run the `listwalletdir` RPC, or File -> Open Wallet in the GUI menu.

Fixes #20081.
Fixes #21136.
Fixes #21904.

Also https://bitcoin.stackexchange.com/questions/99243/listwalletdir-access-is-denied-d-system-volume-information

hebasto committed:
wallet: Do not iterate a directory if having an error while accessing it

This change prevents infinite looping for, for example, system folders
on Windows.

laanwj commented on src/wallet/db.cpp:
```diff
@@ -19,6 +23,12 @@ std::vector<fs::path> ListDatabases(const fs::path& wallet_dir)
         if (ec) {
             LogPrintf("%s: %s %s\n", __func__, ec.message(), it->path().string());
+#ifdef WIN32
+            if (ec.value() == boost::system::windows_error::access_denied) {
```
Is this specific to windows? Or could this happen with directory "access denied" errors on other OSes as well?

hebasto replied:
> Is this specific to windows?

The error code is specific to windows.

> Or could this happen with directory "access denied" errors on other OSes as well?

Yes, it could. But on other OSes the reason of "access denied" errors is a bad setup that a user can/should fix. On windows this situation seems common and not always fixable by a user.

promag replied:
Maybe it should avoid recursion regardless of the error and platform? Is there any case where this approach is not desirable?

hebasto replied:
> Maybe it should avoid recursion regardless of the error and platform? Is there any case where this approach is not desirable?

To be sure I understand you correctly, you are suggesting do not iterate a directory if we have any error while accessing it, right?

promag replied:
Yes, that was my suggestion. Is there any downside of that approach?

hebasto replied:
@laanwj @promag 

Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/21907#issuecomment-839610690).

sipsorcery commented:
Started with:

`bitcoin-qt.exe -walletdir=c:\`

Is this error message expected? The app continues as normal once `OK` is clicked.

![bitcoin-qt_walletdir](https://user-images.githubusercontent.com/197660/117813712-4f788a80-b25b-11eb-9e1d-976f73fa7b5a.png)

On `master` the message is a warning:

![bitcoin_qt_warning](https://user-images.githubusercontent.com/197660/117814047-bac25c80-b25b-11eb-95a8-5b5aa967d1f7.png)

hebasto commented:
> Sure I was only attempting to test the fix in this PR by getting the wallet logic to scan the root drive. I don't have a `D:\` drive.

A USB flash-drive could work, no?

prayank23 reviewed (commented):
1. `bitcoin-qt.exe -walletdir=E:\` Not Responding when I try to open wallets

![image](https://user-images.githubusercontent.com/13405205/117941796-0e9a7780-b328-11eb-8a2e-1b1c2110247a.png)

2. `bitcoind.exe -walletdir=E:\`

Not sure why its using forward slash which is used in Linux
```
2021-05-12T08:05:49Z Using wallet directory E:/
2021-05-12T08:05:49Z init message: Verifying wallet(s)...
2021-05-12T08:05:49Z Using BerkeleyDB version Berkeley DB 4.8.30: (April  9, 2010)
2021-05-12T08:05:49Z Using wallet E:/wallet.dat
```
    
`bitcoin-cli.exe listwalletdir` returns nothing and cmd prompt running bitcoind starts printing error continuously until I close it.
    
![image](https://user-images.githubusercontent.com/13405205/117941987-3d185280-b328-11eb-8ae9-2c8cb710b6ac.png)


Will compile the branch for this PR in sometime and two more things that I noticed:

1. `"E:"` works when used with `datadir` but gives error for `walletdir`
2. https://bitcoin.stackexchange.com/questions/103295/unable-to-use-blocks-and-chainstate-from-external-drive

promag reviewed (commented):
Concept ACK.

hebasto commented:
Updated bd13ec6d834d4c5f20a5ff9ab41e2e9e86f32d93 -> 29c9e2c2d2015ade47ed4497926363dea3f9c59b ([pr21907.01](https://github.com/hebasto/bitcoin/commits/pr21907.01) -> [pr21907.02](https://github.com/hebasto/bitcoin/commits/pr21907.02), [diff](https://github.com/hebasto/bitcoin/compare/pr21907.01..pr21907.02)):

- addressed @laanwj's [comment](https://github.com/bitcoin/bitcoin/pull/21907#discussion_r630076648) and @promag's [comment](https://github.com/bitcoin/bitcoin/pull/21907#discussion_r630836556)

prayank23 commented:
Compiled branch for this PR

1. `bitcoin-qt.exe -walletdir=E:\` I am able to see some names in File->Open Wallet
2. `bitcoind.exe -walletdir=E:\` and `bitcoin-cli.exe listwalletdir` works fine

```
2021-05-12T13:35:53Z ListDatabases: Access is denied E:/System Volume Information -- skipping.
```

I see few names returned by `listwalletdir`

ACK https://github.com/bitcoin/bitcoin/pull/21907/commits/29c9e2c2d2015ade47ed4497926363dea3f9c59b

Found some issues with weird wallet names but its not related to this PR: https://github.com/bitcoin/bitcoin/issues/21510#issuecomment-839803680

promag reviewed (commented):
Code review ACK 29c9e2c2d2015ade47ed4497926363dea3f9c59b.

meshcollider reviewed (commented):
Code review ACK 29c9e2c2d2015ade47ed4497926363dea3f9c59b

fanquake commented:
Backported to 0.21 in #22255.
